core/src/bitcoin_syncer.rs

```diff
@@ -0,0 +1,205 @@
+use bitcoincore_rpc::RpcApi;
+use futures::future::try_join_all;
+use tokio::{sync::broadcast, task::JoinHandle, time::sleep};
+
+use crate::{
+    database::{Database, DatabaseTransaction},
+    errors::BridgeError,
+    extended_rpc::ExtendedRpc,
+};
+
+type ChainHeadPollingResult = (
+    broadcast::Sender<BitcoinSyncerEvent>,
+    JoinHandle<Result<(), BridgeError>>,
+);
+
+#[derive(Clone, Debug)]
+pub struct BlockInfo {
+    pub block_hash: BlockHash,
+    pub block_header: Header,
+    pub block_height: u64,
+}
+
+#[derive(Clone, Debug)]
+pub struct BlockInfoWithTxs {
+    pub block_info: BlockInfo,
+    pub txs: Vec<Transaction>,
+}
+
+#[derive(Clone, Debug)]
+pub enum BitcoinSyncerEvent {
+    NewBlocks(Vec<BlockInfo>),
+    NewBlocksWithTxs(Vec<BlockInfoWithTxs>),
+    ReorgedBlocks(Vec<(BlockHash, BlockHash)>),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum BitcoinSyncerPollingMode {
+    SyncOnly,
+    SyncAndPollTxs,
+}
+
+pub async fn get_block_info_from_height(
+    rpc: &ExtendedRpc,
+    height: u64,
+) -> Result<BlockInfo, BridgeError> {
+    let block_hash = rpc.client.get_block_hash(height).await?;
+    let block_header = rpc.client.get_block_header(&block_hash).await?;
+    Ok(BlockInfo {
+        block_hash,
+        block_header,
+        block_height: height,
+    })
+}
+
+pub async fn process_block(
+    db: &Database,
+    block: &bitcoin::Block,
+    block_height: i64,
+) -> Result<(), BridgeError> {
+    let mut dbtx = db.begin_transaction().await?;
+    let block_hash = block.header.block_hash();
+
+    db.set_chain_head(
+        Some(&mut dbtx),
+        &block_hash,
+        &block.header.prev_blockhash,
+        block_height,
+    )
+    .await?;
+
+    for tx in &block.txdata {
+        process_tx(db, &mut dbtx, tx, &block_hash).await?;
+    }
+
+    dbtx.commit().await?;
+    Ok::<(), BridgeError>(())
+}
+
+pub async fn process_tx(
+    db: &Database,
+    dbtx: DatabaseTransaction<'_, '_>,
+    tx: &bitcoin::Transaction,
+    block_hash: &bitcoin::BlockHash,
+) -> Result<(), BridgeError> {
+    let txid = tx.compute_txid();
+    db.insert_tx(dbtx, block_hash, &txid).await?;
+
+    for input in &tx.input {
+        db.insert_utxo(
+            dbtx,
+            &txid,
+            &input.previous_output.txid,
+            input.previous_output.vout as i64,
+        )
+        .await?;
+    }
+
+    Ok::<(), BridgeError>(())
+}
+
+pub async fn start_bitcoin_syncer(
+    db: Database,
+    rpc: ExtendedRpc,
+    poll_delay: Duration,
+    mode: BitcoinSyncerPollingMode,
+) -> Result<ChainHeadPollingResult, BridgeError> {
+    let (tx, _) = broadcast::channel(100);
+
+    let returned_tx = tx.clone();
+
+    let mut block_height = db
+        .get_max_height(None)
+        .await?
+        .ok_or(BridgeError::BlockNotFound)?;
+
+    let handle = tokio::spawn(async move {
+        loop {
+            let mut block_hash = rpc.client.get_block_hash(block_height + 1).await?;
```

Comment:
Is `?` right here? Wouldn't this return early if there were no block? Instead, I believe we want to wait for poll_delay then go back to the beginning of the loop.